python/hail/expr/expression.py

```diff
@@ -250,6 +255,32 @@ def unify_all(*exprs):
     return new_indices, agg, joins, refs
 
 
+def unify_types(t1, t2):
+    c1 = t1.__class__
+    c2 = t2.__class__
+    if c1 is c2:
+        return t1
+    elif c1 in __numeric_types and c2 in __numeric_types:
+        s = {c1, c2}
+
+        # assert the two classes are different
+        assert len(s) == 2
+        if TFloat64 in s:
+            return TFloat64()
+        elif TFloat32 in s:
+            return TFloat32()
+        else:
+            assert s == {TInt32, TInt64}
+            return TInt64()
+    elif c1 == TArray and c2 == TArray:
```

Comment:
Should this handle the Set case as well?

Reply:
It looks like we don't handle this in expr right now:
```
<input>:1:if (true) [1.0, 2.0].toSet() else [3, 4].toSet()
          ^

Java stack trace:
is.hail.utils.HailException: expected same-type `then' and `else' clause, got `Set[Float64]' and `Set[Int32]'
<input>:1:if (true) [1.0, 2.0].toSet() else [3, 4].toSet()
```

Reply:
actually, this is a problem in ifs too:
```
<input>:1:if (true) [1.0, 2.0] else [3, 4]
          ^

Java stack trace:
is.hail.utils.HailException: expected same-type `then' and `else' clause, got `Array[Float64]' and `Array[Int32]'
<input>:1:if (true) [1.0, 2.0] else [3, 4]
```

Reply:
very good catch! will fix in the scala code for this PR.

Reply:
on second thought, delaying this until later. But am catching the error in python!